### DISCUSSION TITLE
feat:  `exa -l` doesn't show group info (from `exa`)

### DISCUSSION THREAD

**Discussion by mtalexan** [eza-community/eza#199] [Development]
From exa Issue: https://github.com/ogham/exa/issues/1118

tl;dr  
The `ls -l` command includes the name of the group that the "group permissions" apply to.  This is pretty commonly used info, but requires adding the `--group` flag to get it now.  Please add it back to the default, or make it a customizable default for when `-l` is used.

**Comment by mtalexan** [comment]
@ariasuni mentioned in the original Issue report when redirecting here, that they hadn't personally used the group info in the past year. 

As a counter point, I've personally used it almost every day for the past year.  If you're working with device nodes (`/dev`), containers, configuring servers, or debugging problems in newer distros (e.g. immutable distros), a missing group membership is often a consideration/factor.  Many/most system tools use group ownership of files, folders, sockets, and devices to manage which users have access to use the tools.  Installing those system tools, debugging them, or trying to use them within a container means regularly verifying the group on them and comparing it to the groups a user is a member of (especially in containers where the GID may need to be explicitly mapped).  
In the context of file ownership, my experience has been that it's only a little less likely you need to know the group as that you need to know the owner of a file.  

So the things I personally use the extended info for, in order of frequency, are:
1. Is it a symlink, folder, file, block device, or character device?
2. Where does the symlink point?
3. What are the `rwx` permissions on the file?
4. Who's the owner of the file?
5. What's the group of the file?
6. How big is the file?
7. Are there ACL permissions?
8. When was the file last modified?
9. What's the major/minor number of the device node?

If you notice, the last modified time is actually one of the least likely things I'm looking for, though I definitely understand that many people use that info a lot more often than I do.  Similarly I understand that most people probably check the size of the file way more often than they consider the group.  

---

I think it makes a lot of sense to match the native `ls -l` behavior in terms of showing the group of a file, the `rwx` permissions are grouped into 3 sets for `user/owner`, `group`, and `other`, so it's part of the permissions description to know which group that `group` permission applies to.

**Comment by cafkafk** [comment]
> From exa Issue: [ogham/exa#1118](https://github.com/ogham/exa/issues/1118)
> 
> tl;dr The `ls -l` command includes the name of the group that the "group permissions" apply to. This is pretty commonly used info, but requires adding the `--group` flag to get it now. Please add it back to the default, or make it a customizable default for when `-l` is used.

We might make it possible to set a default whenever I have time to look into #139, but until then, perhaps an alias would be sufficient. We could still consider changing the default behavior, but this would probably be better handled by a discussion thread than an issue.

I personally use an alias to get the default eza behavior I need, and I'm cautious about enforcing my overly opinionated takes onto the default user.

That said, I do see where you're coming from, but I'd like to see a wider discussion before moving forward.

**Comment by ferki** [comment]
I, for one, also often miss group ownership info when using `-l`, especially given that the permission info is fully printed — I'm told what the group members may or may not do, but I have no clue which group is that :shrug:

That feels inconsistent as user experience.

**Comment by tcurdt** [comment]
I am OKish not showing the group if it's the same as the user but this here is just a pretty bad UX:

```
# eza -l
.rw-r----- 128 root 10 Mar 15:30 foo.env
# eza -lg
.rw-r----- 128 root foo 10 Mar 15:30 foo.env
```

If user != group it should never be hidden (unless I asked it to).

**Reply by Omnikron13** [reply]
Most of the time it is tbh nice that the 'stuttering' username/group isn't there, with _most_ files having the same user and group.

The real question is; what is the desired behaviour? Suppress the group column still if all the files being printed _are_ the same group & user? And then if a single file isn't, trip displaying the column for all files so there's loads of `foo foo` in the output? Or some nicer way of indicating that the user & group match for most files, while still printing them separately where they differ? (the latter might actually be pretty useful tbh to make them stand out more)

**Reply by tcurdt** [reply]
Ignoring the opinion part of that comment, I would argue that `ls -la` has always been the "show me the details" command. And right now that's missing something.

I see 4 different possible display modes:

1. none
2. username
3. username + group
4. username + difference indicator

Question is what should be the default and what should be supported.

I personally think 3. should be the default.
I can see 4. making sense for a non-OPS person.
But 2. feels like very bad idea to me. Especially as a default.

**Reply by Omnikron13** [reply]
I think that's the crux of the issue here really; number 2 is actually ideal _most_ of the time because it just prevents `alice alice` or w/e cluttering the output, but when there actually _is_ a difference you're potentially much more likely to be interested in knowing about it.

I think potentially there's a possibility of actually improving on the output of `ls -l` by printing out `alice ???` (I'm not sure what exactly) instead of `alice alice`, and then the `alice aruba` listing (or w/e) stands out even more.

Would of course require a little examination of the files to see if the group column is actually adding any information before printing, but I'd say that's worth it to provide more information at a glance to the user imo.

**Reply by tcurdt** [reply]
Maybe we can remove the bias from the discussion?

I already understood, that you think 2. is ideal - for you.
I'd challenge the "most of the time" and very much disagree.

It seems like 4 could be a compromise:

`username ` could be `username username`
`username*` could be `username groupname-is-different-from-username`

But even that would not make a sensible default for an OPS minded person.
So I guess the background of the user matters.

**Reply by Omnikron13** [reply]
I'm perhaps not expression what I mean very well... I specifically mean that if every entry that is being printed out has identical user and group data, then it's superfluous to actually print them both out. I agree it's clearly an unintuitive and misleading problem if any files in the output _aren't_  identical, and the extra column really ought to be there as would be expected coming from `ls`.
 
A more elegant and readable solution might be to either use colour, an icon, something like `username*         `  when they _do_ match, or to fill the empty column in some way like `steven [....]` perhaps? Unless the intention is process the output in some way this is perhaps an issue, though I've been operating under the assumption that this is not the primary use case of eza.

If this was the behaviour it would be pretty quick and easy to see if any non-matching files were output, and to find them in the list as they would have properly filled columns there.

> I'd challenge the "most of the time" and very much disagree.

This getting a little tangential really from the possible solutions here, but I suppose we can't really expect some degree of bias to not creep into such things, and can't be unbiased without looking at how frequently directories contain files with non-matching username/group children, and how often people are using the long output in those directories vs. others.

**Reply by mtalexan** [reply]
It seems a lot of thought and work is going into avoiding printing both the user and group columns when they're the same name.
It should probably be pointed out that only about half of Linux distros assign a matching group name, and only to locally created users at that. If you have an SSO/LDAP/Active Directory user, they're usually assigned to a fixed group (i.e. `domain-users`). If you have a server version of basically any distros, it also assigns locally created users to a fixed group (i.e. `users`). If you're running a desktop version of one of the other 50% of distros, it also assigns local users to a fixed group (i.e. `users`).

So a lot of time and energy is being put into avoiding showing a duplicate of the name when only Desktop installs of some distros for a local user even fit that case.

**Reply by tcurdt** [reply]
>  I suppose we can't really expect some degree of bias

Agreed - but then it should maybe not act as the main argument. It's better to concentrate on more tangible points.

This is problematic from a UX POV (and IMO should **not** be the default)
```
drwxr-xr-x     - tcurdt 2024-01-09 18:33 shells
drwxr-xr-x     - tcurdt 2024-03-09 13:33 users
```

This gives the expected information (but some -not me- might consider it verbose)
```
drwxr-xr-x     - tcurdt staff  2024-01-09 18:33 shells
drwxr-xr-x     - tcurdt tcurdt 2024-03-09 13:33 users
```

This might be a middle ground:
```
drwxr-xr-x     - tcurdt* 2024-01-09 18:33 shells
drwxr-xr-x     - tcurdt  2024-03-09 13:33 users
```

**Comment by Omnikron13** [comment]
> This is problematic from a UX POV (and IMO should **not** be the default)
> 
> ```
> drwxr-xr-x     - tcurdt 2024-01-09 18:33 shells
> drwxr-xr-x     - tcurdt 2024-03-09 13:33 users
> ```
Yeah this is really not acceptable for in the case of this data set.

> This gives the expected information (but some -not me- might consider it verbose)
> 
> ```
> drwxr-xr-x     - tcurdt staff  2024-01-09 18:33 shells
> drwxr-xr-x     - tcurdt tcurdt 2024-03-09 13:33 users
> ```
Well, it is certainly displaying rather redundant information that could be inferred, especially compared to this counter example date, where actual variances may get lost in the noise.
```
drwxr-xr-x     - tcurdt tcurdt  2024-01-09 18:33 shells
drwxr-xr-x     - tcurdt tcurdt  2024-03-09 13:33 users
drwxr-xr-x     - tcurdt tcurdt  2024-05-09 08:33 other
drwxr-xr-x     - tcurdt tcurdt  2024-05-09 03:33 things
```

> This might be a middle ground:
> 
> ```
> drwxr-xr-x     - tcurdt* 2024-01-09 18:33 shells
> drwxr-xr-x     - tcurdt  2024-03-09 13:33 users
> ```

This would definitely display a _bit_ more information, but not really the specifics you might want to know;...

Un-hiding to the column but indicating it contains the same data would seem to be more conducive to quickly reading the output. Something more like this:

```
drwxr-xr-x     - tcurdt staff   2024-01-09 18:33 shells
drwxr-xr-x     - tcurdt   ~     2024-03-09 13:33 users
drwxr-xr-x     - tcurdt   ~     2024-05-09 08:33 other
drwxr-xr-x     - tcurdt   ~     2024-05-09 03:33 things
```

(The actual character, or icon, or characters, colour, etc. standing in for the repetition would have to have more thought put into it ofc. Or be declared by themes tbh.)

**Reply by tcurdt** [reply]
If keeping the column is fine, I don't really get the point of replacing the group with a "same" indicator.
I'd be OK with it - but I don't think it really helps. I would just show the group by default and be done with it.

**Comment by rlkiser** [comment]
New eza user. Sorry if I'm resurrecting this, but I Just found the issue/discussion and wanted to add that it's definitely thrown me off a couple times not having group in -l results by default. An alias is an easy fix, but it IS something I found unusual fwiw. 

It's also exactly the kind of little thing I can definitely see myself forgetting to set in a new build, then working through troubleshooting the issue until I remember.

.02c

**Comment by cubeleo** [comment]
Another +1 wishing for default -l behavior for group.  I'm working with /dev files and needed to know the group.  I'm working on a cross platform code base, so weird subtleties like this hurt even more when long standing conventions are weirdly missing as I bounce between platforms and build environments.  In fact this issue is what caused me to realize that eza exists since ls is aliased to eza on CachyOS which I've been playing with recently.  After lots of typing ls -l and wondering what happened to my reality I dug deeper and figured it out.

**Comment by Desalutar20** [comment]
Is this will be changed or no ?
